website_sale_coupon_page/controllers/main.py

```diff
@@ -0,0 +1,27 @@
+# Copyright 2021 Tecnativa - Carlos Roca
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+from odoo import http
+from odoo.http import request
+
+
+class WebsiteSale(http.Controller):
+    @http.route(["""/promotions"""], type="http", auth="public", website=True)
+    def promotion(self, **post):
+        all_promos = (
+            request.env["sale.coupon.program"]
+            .sudo()
```

Comment:
```suggestion
            .with_user(request.env.user)
```

Reply:
Public or portal users don't have access to this model